### DISCUSSION TITLE
What is the meaning of "currentValue" in renderLoop(unsigned int currentValue, RenderPressMode userClick)

### DISCUSSION THREAD

**Discussion by luwenjin** [davetcc/tcMenu#295] [API & Remote]
It seems that the `currentValue` of `renderLoop(unsigned int currentValue, RenderPressMode userClick)` shows the index of the active item which was hidden by `renderer.takeOverDisplay();`, and the max value of `currentValue` is restricted by the menu item count.

IMO, since the display is already taken by custom ui, `currentValue` should not relate to the items of the current menu.

**Comment by davetcc** [comment]
I've moved this to a discussion because it's more of a question than an issue.

The current value is nothing more than a convenient way to get the current value of the rotary encoder in the render loop. You can change the precision of the encoder when starting the takeover of the display to whatever range is needed, the menu manager will automatically set it back upon giving the display back.

For example in an amplifier, a dashboard showing volume and programme information may be shown, turning the encoder in that case would adjust the volume. Clicking the button return to the main menu.

In the early days before I provided this, many people who just wanted to use the encoders value within the takeover display callback had to manually register their own encoder, so that they could get hold of the value from it, these days it is less of an issue because the encoder has a function to get the current value anyway. But even so it is still a convenience for many cases.

The user click is even less clear, it is only usable for the case where you want to detect the click to return to the menu. As once it goes to clicked it remains clicked until the display is given back.

**Reply by luwenjin** [reply]
> turning the encoder in that case would adjust the volume. Clicking the button return to the main menu.

I suppose UP/DOWN buttons work the same as encoders, and I found 3 problems with UP/DOWN buttons: 

1. The max of `currentValue` is limited by the number of menu items. 
  If I `takeoverdisplay` under a menu with 3 items, the max value of `currentValue` will be only 3. 
  if I `takeoverdisplay` by enter the 3rd item, the default value of `currentValue` will be 3, too.
  So I would not be able to make an amplifier with 256 levels there.

2. the type of 'currentValue' is unsigned int, by design It would be difficult to make an "amplifier" with both positive and negative values.

3. I've tried to hold the OK button for very long time, but the `userClick` never become `RPRESS_HELD = 2`. Actually I don't need RPRESS_HELD, but since it's an legit RenderPressMode, I tried that, but have no luck.

**Reply by davetcc** [reply]
1. You’ve misunderstood, please look at the various examples that take over the display. They use the switches object and change the range the encoder represents. 

2. The API.can’t really change as it is too baked in. Whatever we have is set in stone. In reality it is from 0 to the precision you set. 

3.  Could well be a bug but would need to check.

**Comment by luwenjin** [comment]
Thanks for the suggestion.
I've tried the switches object and changed precision of the encoder. It worked perfect!
